aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
-  } else if (should_fold_into_mm(tensor1, tensor2)) {
+  } else if (should_fold_into_mm(tensor1, tensor2) ||
+             should_fold_into_mm(tensor2, tensor1)) {
     // optimization: use mm instead of bmm by folding tensor1's batch into
-    // its leading matrix dimension.
-
-    Tensor t2 = dim_tensor2 == 1 ? tensor2.unsqueeze(-1) : tensor2;
-    auto size1 = tensor1.sizes();
-    auto size2 = t2.sizes();
-    DimVector output_size(size1.begin(), size1.end() - 1);
-    if (dim_tensor2 > 1) {
-      output_size.push_back(size2[dim_tensor2 - 1]);
-    }
-
-    // fold the batch into the first dimension
-    // Why not tensor1.view(-1, size1[size1.size() -1])?
+    // its leading matrix dimension
+    // dim_tensor1 >=3 && (dim_tensor2 == 1 || dim_tensor2 == 2) ||
+    // dim_tensor2 >=3 && (dim_tensor1 == 1 || dim_tensor1 == 2)
+    // and some condition on the strides is fulfilled
+    const auto transpose = dim_tensor2 > dim_tensor1;
+    const auto t1 = transpose ? MaybeOwned<Tensor>::owned(tensor2.transpose(-2, -1))
+                              : MaybeOwned<Tensor>::borrowed(tensor1);
+    const auto t2 = transpose && dim_tensor1 == 2
+                      ? MaybeOwned<Tensor>::owned(tensor1.t())
+                      : MaybeOwned<Tensor>::borrowed(transpose ? tensor1 : tensor2);
+    // Invariant: t1->dim() >= 3 && (t2->dim() == 1 || t2->dim() == 2)
+
+    // Why not t1->view(-1, sizes_1.back())?
     // If the last dim is 0, then view(-1, 0) won't work because the -1 becomes ambiguous.
     // This can happen in e.g. [3, 5, 0] @ [0, 0].
-    // So we manually compute the folding as a result.
-    const auto dim1_size = c10::multiply_integers(size1.begin(), size1.end() - 1);
-    auto t1 = tensor1.expect_contiguous()->view({dim1_size, size1[size1.size() - 1]});
-    Tensor output = has_out ? at::_unsafe_view(at::mm_out(out, t1, t2), output_size)
-                            : at::_unsafe_view(t1.mm(t2), output_size);
-    return has_out ? out.set_(output) : output;
-  } else if ((dim_tensor1 == 1 || dim_tensor1 == 2) && dim_tensor2 >= 3) {
-    // optimization: transpose the inner dimensions of the arguments, call
-    // matmul on the swapped arguments, then transpose the inner dimensions
-    // of the result.
-    const int64_t n = dim_tensor1 == 2 ? tensor1.size(-2) : 1;
-    const int64_t m = tensor1.size(-1);
-    const int64_t p = tensor2.size(-1);
-
-    const Tensor t2_T = tensor2.transpose(-1, -2);
-    const Tensor t1_T = dim_tensor1 == 2 ? tensor1.t() : tensor1.reshape({n, m}).t();
-    const Tensor res_T = _matmul_impl(out, t2_T, t1_T);
-
-    if (dim_tensor1 == 2) {
-      Tensor res = res_T.transpose(-1, -2).contiguous();
-      return has_out ? out.set_(res) : res;
+    const auto sizes_1 = t1->sizes();
+    auto output_shape = DimVector(sizes_1.begin(), sizes_1.end() - 1);
+    const auto batch_product = c10::multiply_integers(output_shape);
```

Comment:
It's probably worth pointing out that batch product here has ALREADY done the folding (otherwise this feels like an off by one error)